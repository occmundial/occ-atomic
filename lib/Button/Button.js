import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import useStyles from './styles';

/**
* A flexible button component, with four pre-stablished themes.
* The Buton can behave like any button with an onClick function, or like an anchor with href and target.
* There are 3 sizes available: small, medium and large.
*/
const Button = props => {
    const { className, children, id, href, target, rel, onClick, theme, size, block, disabled, icon, iconRight, style, loading } = props;
    const classes = useStyles(props);
    const buttonClassName = classnames(
        classes.btn,
        { [classes[theme]]: theme },
        { [classes.loading]: loading },
        { [classes.disabled]: disabled },
        { [classes[size]]: ['md', 'lg'].includes(size) },
        { [classes.block]: block },
        className,
        { [classes.iconOnly]: !children && icon }
    );
    const content = (
        <span className={classes.cont}>
            {icon ? <i className={classes.icon} /> : ''}
            {children}
            {iconRight ? <i className={classes.iconRight} /> : ''}
        </span>
    );
    const loadingLayer = loading ? (
        <span className={classes.loadCont}>
            <i className={classes.loadIcon} />
        </span>
    ) : null;
    if (href) {
        return (
            <a
                className={buttonClassName}
                href={!disabled ? href : ''}
                target={target}
                rel={rel}
                onClick={!disabled ? onClick : null}
                id={id}
                style={style}>
                {content}
                {loadingLayer}
            </a>
        );
    } else {
        return (
            <button
                className={buttonClassName}
                onClick={!disabled ? onClick : null}
                id={id}
                style={style}>
                {content}
                {loadingLayer}
            </button>
        );
    }
};

Button.defaultProps = {
    theme: 'primary',
    size: 'sm',
    block: false,
    disabled: false
};

Button.propTypes = {
    /** There are four themes available: main, secondary, white, blue, flat. */
    theme: PropTypes.oneOf([
        'primary', 'secondary', 'tertiary', 'tertiaryWhite', 'ghostPink', 'ghostGrey', 'ghostWhite'
    ]),
    /** The button has three different sizes: sm, md, lg. */
    size: PropTypes.oneOf([
        'sm', 'md', 'lg'
    ]),
    /** The block property lets the button expand to ocupy the whole width of its parent. */
    block: PropTypes.bool,
    /** It disables the button and shows it with the proper theme. */
    disabled: PropTypes.bool,
    /** Use the name of any of the icons available in the library. */
    icon: PropTypes.string,
    /** Use the name of any of the icons available in the library. The icon will be rendered on the right side of the button. */
    iconRight: PropTypes.string,
    /** The loading property disables the button and shows a loading icon spinning. */
    loading: PropTypes.bool,
    /** The function to call when the button is clicked. */
    onClick: PropTypes.func,
    /** Optionally, you can set the href property and use it like a link. */
    href: PropTypes.string,
    /** If you use the href property, you can also define the target. */
    target: PropTypes.string,
    /** If you use the href property, you can also define the rel attribute. */
    rel: PropTypes.string,
    id: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object
};

export default Button;
