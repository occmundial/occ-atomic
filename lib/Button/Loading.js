import React from 'react';
import PropTypes from 'prop-types';

export default function Loading(props) {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      viewBox="0 0 24 24"
      {...props}
    >
      <circle cx="4" cy="13" r="3">
        <animate
          attributeName="r"
          begin="0s"
          dur="0.7s"
          values="2;3;2"
          calcMode="linear"
          repeatCount="indefinite"
        />
        <animate
          attributeName="cy"
          begin="0s"
          dur="0.7s"
          values="10;14;10"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="12" cy="13" r="3">
        <animate
          attributeName="r"
          begin="0.1s"
          dur="0.7s"
          values="2;3;2"
          calcMode="linear"
          repeatCount="indefinite"
        />
        <animate
          attributeName="cy"
          begin="0.1s"
          dur="0.7s"
          values="10;14;10"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="20" cy="13" r="3">
        <animate
          attributeName="r"
          begin="0.2s"
          dur="0.7s"
          values="2;3;2"
          calcMode="linear"
          repeatCount="indefinite"
        />
        <animate
          attributeName="cy"
          begin="0.2s"
          dur="0.7s"
          values="10;14;10"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
    </svg>
  );
};

Loading.propTypes = {
    fill: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
};