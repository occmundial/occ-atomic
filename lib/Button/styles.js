import legacyColors from '../subatomic/colors';
import legacyFont from '../subatomic/fonts';
import icons from '../subatomic/icons';

import colors from '../tokens/colors.json';
import spacing from '../tokens/spacing.json';
import borderRadius from '../tokens/borderRadius.json';
import fonts from '../tokens/fonts.json';
import shadows from '../tokens/shadows.json';
import { legacyIcons } from '../subatomic/mappedIcons';

const { sec, secDark, grey500, secLight } = legacyColors;

const sm = fonts['button-small'];
const md = fonts['button-medium'];
const lg = fonts['button-large'];

const renderIcon = (props, color) => {
    if (!legacyIcons[props.icon]) return '';
    if (props.icon) {
        return icons.base(icons[props.icon].icon(color));
    } else if (props.iconRight) {
        return icons.base(icons[props.iconRight].icon(color));
    }
    return '';
};

export default {
    btn: {
        display: 'inline-block',
        boxSizing: 'border-box',
        position: 'relative',
        maxWidth: '100%',
        marginBottom: 0,
        border: 0,
        padding: [spacing['size-3'], spacing['size-4']],
        borderRadius: borderRadius['br-xs'],
        font: `${sm.fontWeight} ${sm.fontSize}/${sm.lineHeight} ${legacyFont.body}`,
        textAlign: 'center',
        textDecoration: 'none',
        whiteSpace: 'nowrap',
        transition: '0.3s all',
        cursor: 'pointer',
        userSelect: 'none',
        touchAction: 'manipulation',
        '&:focus': {
            outline: 'none'
        },
        '&:hover': {
            textDecoration: 'none'
        }
    },
    cont: {
        display: 'flex',
        width: '100%',
        height: '100%',
        alignItems: 'center',
        justifyContent: 'center',
        pointerEvents: 'none'
    },
    loadCont: {
        display: 'flex',
        width: '100%',
        height: '100%',
        alignItems: 'center',
        justifyContent: 'center',
        position: 'absolute',
        top: 0,
        left: 0,
        pointerEvents: 'none'
    },
    // Themes
    primary: {
        background: colors.button.primary.bg.default,
        color: colors.text.white.primary,
        outline: `2px solid ${colors.button.primary.border.default}`,
        outlineOffset: '-2px',
        '&:hover': {
            background: colors.button.primary.bg.hover
        },
        '&:active': {
            background: colors.button.primary.bg.active
        },
        '&:focus': {
            outline: `2px solid ${colors.button.primary.border.default}`
        },
        '&:not(:active):focus-visible': {
            boxShadow: shadows['focus-pink']
        },
        '&$disabled': {
            background: colors.button.primary.bg.disabled,
            color: colors.text.white.secondary,
            cursor: 'not-allowed'
        },
        '& span span': {
            background: props => renderIcon(props, [colors.icon.inverse.default])
        },
        '&$disabled span span': {
            background: props => renderIcon(props, [colors.icon.inverse.disabled])
        }
    },
    secondary: {
        background: colors.button.secondary.bg.default,
        color: colors.text.indigo.primary,
        '&:hover': {
            background: colors.button.secondary.bg.hover
        },
        '&:active': {
            background: colors.button.secondary.bg.active
        },
        '&:not(:active):focus-visible': {
            boxShadow: shadows['focus-indigo']
        },
        '&$disabled': {
            background: colors.button.secondary.bg.disabled,
            color: colors.text.indigo.secondary
        },
        '& span span': {
            background: props => renderIcon(props, [colors.icon.brand.default])
        },
        '&$disabled span span': {
            background: props => renderIcon(props, [colors.icon.brand.disabled])
        },
        '& $loadIcon': {
            background: icons.base(icons.loading.icon([colors.icon.brand.default]))
        }
    },
    tertiary: {
        background: 'transparent',
        color: colors.text.indigo.primary,
        outline: `2px solid ${colors.button.tertiary.border.default}`,
        outlineOffset: '-2px',
        '&:hover': {
            background: colors.button.tertiary.bg.hover
        },
        '&:active': {
            background: colors.button.tertiary.bg.active
        },
        '&:focus': {
            outline: `2px solid ${colors.button.tertiary.border.default}`
        },
        '&:not(:active):focus-visible': {
            boxShadow: shadows['focus-indigo']
        },
        '& span span': {
            background: props => renderIcon(props, [colors.icon.brand.default])
        },
        '&$disabled': {
            background: 'transparent',
            color: colors.text.indigo.secondary,
            outline: `2px solid ${colors.button.tertiary.border.disabled}`
        },
        '&$disabled span span': {
            background: props => renderIcon(props, [colors.icon.brand.disabled])
        },
        '& $loadIcon': {
            background: icons.base(icons.loading.icon([colors.icon.brand.default]))
        }
    },
    tertiaryWhite: {
        background: 'transparent',
        color: colors.text.white.primary,
        outline: `2px solid ${colors.button.tertiary.border.inverse.default}`,
        outlineOffset: '-2px',
        '&:hover': {
            background: colors.button.tertiary.bg.inverse.hover
        },
        '&:active': {
            background: colors.button.tertiary.bg.inverse.active
        },
        '&:focus': {
            outline: `2px solid ${colors.button.tertiary.border.inverse.default}`
        },
        '&:not(:active):focus-visible': {
            boxShadow: shadows['focus-white']
        },
        '& span span': {
            background: props => renderIcon(props, [colors.icon.inverse.default])
        },
        '&$disabled': {
            background: 'transparent',
            color: colors.text.white.secondary
        },
        '&$disabled span span': {
            background: props => renderIcon(props, [colors.icon.inverse.disabled])
        }
    },
    ghostPink: {
        background: 'transparent',
        color: sec,
        border: 'none',
        paddingLeft: [0, '!important'],
        paddingRight: [0, '!important'],
        '&:hover, &:active': {
            color: secDark,
            '& span span': {
                background: props => renderIcon(props, [secDark])
            },
        },
        '& span span': {
            background: props => renderIcon(props, [sec])
        },
        '&$disabled': {
            background: `transparent !important`,
            color: `${secLight} !important`,
            border: `none !important`,
            '& span span': {
                background: props => renderIcon(props, [grey500])
            }
        },
        '& $loadIcon': {
            background: icons.base(icons.loading.icon([sec]))
        }
    },
    ghostGrey: {
        background: 'transparent',
        color: colors.text.corp.secondary,
        '&:hover': {
            background: colors.button.ghost.bg.hover
        },
        '&:active': {
            background: colors.button.ghost.bg.active
        },
        '&:not(:active):focus-visible': {
            boxShadow: shadows['focus-corp']
        },
        '& span span': {
            background: props => renderIcon(props, [colors.icon.default.default])
        },
        '&$disabled': {
            background: 'transparent',
            color: colors.text.corp.disabled
        },
        '&$disabled span span': {
            background: props => renderIcon(props, [colors.icon.default.disabled])
        },
        '& $loadIcon': {
            background: icons.base(icons.loading.icon([colors.icon.default.default]))
        }
    },
    ghostWhite: {
        background: 'transparent',
        color: colors.text.white.primary,
        '&:hover': {
            background: colors.button.ghost.bg.inverse.hover
        },
        '&:active': {
            background: colors.button.ghost.bg.inverse.active
        },
        '&:not(:active):focus-visible': {
            boxShadow: shadows['focus-white']
        },
        '& span span': {
            background: props => renderIcon(props, [colors.icon.inverse.default])
        },
        '&$disabled': {
            background: 'transparent',
            color: colors.text.white.secondary
        },
        '&$disabled span span': {
            background: props => renderIcon(props, [colors.icon.inverse.disabled])
        }
    },
    loading: {
        pointerEvents: 'none',
        '& $cont': {
            visibility: 'hidden'
        }
    },
    disabled: {
        cursor: 'default',
        pointerEvents: 'none',
    },
    // Sizes
    md: {
        font: `${md.fontWeight} ${md.fontSize}/${md.lineHeight} ${legacyFont.body}`,
        padding: [spacing['size-3'], spacing['size-5']],
        '&$iconOnly': {
            padding: [spacing['size-2'], spacing['size-2']]
        }
    },
    lg: {
        font: `${lg.fontWeight} ${lg.fontSize}/${lg.lineHeight} ${legacyFont.body}`,
        padding: [spacing['size-3'], spacing['size-6']],
        '&$iconOnly': {
            padding: [spacing['size-3'], spacing['size-3']]
        }
    },
    // Block
    block: {
        display: 'block',
        width: '100%'
    },
    // Icon
    icon: {
        marginRight: spacing['size-2'],
        display: 'inline-block',
        transition: '0.3s all'
    },
    iconRight: {
        width: 16,
        height: 16,
        marginLeft: spacing['size-2'],
        display: 'inline-block',
        transition: '0.3s all'
    },
    iconOnly: {
        padding: [spacing['size-2'], spacing['size-2']],
        '& span span, & svg': {
            marginRight: 0
        }
    },
    loadIcon: {
        width: 24,
        height: 24,
        fill: colors.icon.inverse.default
    },
    round: {
        borderRadius: '50%'
    }
};
