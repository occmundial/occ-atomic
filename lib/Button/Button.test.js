import React from 'react';
import { mount } from 'enzyme';
import Button from './Button';

describe('Button', () => {

    it('matches the snapshot', () => {
        const wrapper = mount(
            <Button>Button</Button>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders the children content', () => {
        const wrapper = mount(
            <Button>Button</Button>
        );
        expect(wrapper.text()).toEqual('Button');
    });

    it('returns a button', () => {
        const wrapper = mount(
            <Button>Button</Button>
        );
        expect(wrapper.find('button').length).toBe(1);
        expect(wrapper.find('a').length).toBe(0);
    });

    it('returns an anchor', () => {
        const wrapper = mount(
            <Button href="http://occ.com.mx">Button</Button>
        );
        expect(wrapper.find('button').length).toBe(0);
        expect(wrapper.find('a').length).toBe(1);
    });

    it('has the right theme', () => {
        let wrapper = mount(
            <Button>Button</Button>
        );
        expect(wrapper.find('button').prop('className').indexOf('primary')).not.toBe(-1);
        wrapper.setProps({ theme: 'secondary' });
        expect(wrapper.find('button').prop('className').indexOf('primary')).toBe(-1);
        expect(wrapper.find('button').prop('className').indexOf('secondary')).not.toBe(-1);
    });

    it('has the right size', () => {
        const wrapper = mount(
            <Button size="md">Button</Button>
        );
        expect(wrapper.find('button').prop('className').indexOf('md')).not.toBe(-1);
        expect(wrapper.find('button').prop('className').indexOf('lg')).toBe(-1);
        wrapper.setProps({size:'lg'});
        expect(wrapper.find('button').prop('className').indexOf('md')).toBe(-1);
        expect(wrapper.find('button').prop('className').indexOf('lg')).not.toBe(-1);
    });

    it('has the block class', () => {
        const wrapper = mount(
            <Button>Button</Button>
        );
        expect(wrapper.find('button').prop('className').indexOf('block')).toBe(-1);
        wrapper.setProps({block:true});
        expect(wrapper.find('button').prop('className').indexOf('block')).not.toBe(-1);
    });

    it('calls onClick function', () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <Button onClick={onClick} disabled>Button</Button>
        );
        wrapper.simulate('click');
        expect(onClick.mock.calls.length).toBe(0);
        wrapper.setProps({disabled: false});
        wrapper.simulate('click');
        expect(onClick.mock.calls.length).toBe(1);
    });

    it('disabled', () => {
        const wrapper = mount(
            <Button disabled>Button</Button>
        );
        expect(wrapper.find('button').prop('className').indexOf('disabled')).not.toBe(-1);
    });

    it('has an icon', () => {
        const wrapper = mount(
            <Button icon="plus">Button</Button>
        );
        expect(wrapper.find('button span i').prop('className').indexOf('icon')).not.toBe(-1);
    });

    it('has an icon on the right', () => {
        const wrapper = mount(
            <Button iconRight="plus">Button</Button>
        );
        expect(wrapper.find('button span i').prop('className').indexOf('iconRight')).not.toBe(-1);
    });

    it('has a loaging icon', () => {
        const wrapper = mount(
            <Button loading>Button</Button>
        );
        expect(wrapper.find('button span i').prop('className').indexOf('load')).not.toBe(-1);
    });

});
