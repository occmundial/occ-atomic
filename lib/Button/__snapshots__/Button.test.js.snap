// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches the snapshot 1`] = `
ReactWrapper {
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "block": false,
      "children": "Button",
      "disabled": false,
      "size": "sm",
      "theme": "primary",
    },
    "ref": null,
    "rendered": Object {
      "instance": <button
        class="btn-0-2-1 primary-0-2-4 primary-d0-0-2-21"
      >
        <span
          class="cont-0-2-2"
        >
          
          Button
          
        </span>
      </button>,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="cont-0-2-2"
          >
            
            Button
            
          </span>,
          null,
        ],
        "className": "btn-0-2-1 primary-0-2-4 primary-d0-0-2-21",
        "id": undefined,
        "onClick": undefined,
        "style": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": <span
            class="cont-0-2-2"
          >
            
            Button
            
          </span>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "",
              "Button",
              "",
            ],
            "className": "cont-0-2-2",
          },
          "ref": null,
          "rendered": Array [
            "",
            "Button",
            "",
          ],
          "type": "span",
        },
      ],
      "type": "button",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "block": false,
        "children": "Button",
        "disabled": false,
        "size": "sm",
        "theme": "primary",
      },
      "ref": null,
      "rendered": Object {
        "instance": <button
          class="btn-0-2-1 primary-0-2-4 primary-d0-0-2-21"
        >
          <span
            class="cont-0-2-2"
          >
            
            Button
            
          </span>
        </button>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="cont-0-2-2"
            >
              
              Button
              
            </span>,
            null,
          ],
          "className": "btn-0-2-1 primary-0-2-4 primary-d0-0-2-21",
          "id": undefined,
          "onClick": undefined,
          "style": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": <span
              class="cont-0-2-2"
            >
              
              Button
              
            </span>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "",
                "Button",
                "",
              ],
              "className": "cont-0-2-2",
            },
            "ref": null,
            "rendered": Array [
              "",
              "Button",
              "",
            ],
            "type": "span",
          },
        ],
        "type": "button",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__linkedRoots__): Array [],
  Symbol(enzyme.__unrendered__): <Button
    block={false}
    disabled={false}
    size="sm"
    theme="primary"
  >
    Button
  </Button>,
  Symbol(enzyme.__updatedBy__): null,
}
`;
